File: sdk/typescript/src/wasi/io.ts
// @waeasi/sdk — wasi:io/streams facade
//
// Two thin async interfaces over WASI 0.2 streams.  `read` returns
// `null` when the host signals end-of-stream; `writeAll` forces a
// blocking-write-and-flush, which on WAeasi's reactor turns into a
// sched::reactor::Waker registration rather than a busy loop.
//
// Implementation note: jco's generated bindings expose the underlying
// poll/wait machinery as Promises through the asyncify wrapper, so
// `await` in user code Just Works.  In the dev-mode polyfill we
// short-circuit to in-memory buffers.

export interface InputStream {
    /**
     * Read up to `n` bytes.  Returns the chunk read, or `null` on EOF.
     * May return a 0-byte chunk if the host signals "would block but
     * not closed" — caller should `await` the next tick.
     */
    read(n: number): Promise<Uint8Array | null>;
}

export interface OutputStream {
    /** Write the buffer in full; resolves when the host has accepted it. */
    writeAll(chunk: Uint8Array): Promise<void>;
    /** Close the stream — sends FIN / chunked-trailer / wsi:io drop. */
    close(): void;
}

/** Build an InputStream backed by a concrete Uint8Array (tests, fixtures). */
export function bufferStream(data: Uint8Array): InputStream {
    let off = 0;
    return {
        async read(n: number): Promise<Uint8Array | null> {
            if (off >= data.byteLength) return null;
            const end = Math.min(off + n, data.byteLength);
            const slice = data.subarray(off, end);
            off = end;
            return slice;
        },
    };
}

/** Build an OutputStream that accumulates in-memory.  Test helper. */
export function captureStream(): OutputStream & { drain(): Uint8Array } {
    const chunks: Uint8Array[] = [];
    let total = 0;
    let closed = false;
    return {
        async writeAll(c: Uint8Array): Promise<void> {
            if (closed) throw new Error("stream closed");
            chunks.push(c);
            total += c.byteLength;
        },
        close(): void { closed = true; },
        drain(): Uint8Array {
            const out = new Uint8Array(total);
            let off = 0;
            for (const c of chunks) { out.set(c, off); off += c.byteLength; }
            return out;
        },
    };
}

/** Pipe rx → tx, returning total bytes transferred. */
export async function pipe(
    rx: InputStream,
    tx: OutputStream,
    chunkSize = 8 * 1024,
): Promise<number> {
    let total = 0;
    for (;;) {
        const c = await rx.read(chunkSize);
        if (c === null) break;
        if (c.byteLength === 0) {
            // host hint: yield then retry
            await new Promise<void>((res) => queueMicrotask(res));
            continue;
        }
        await tx.writeAll(c);
        total += c.byteLength;
    }
    return total;
}

File: sdk/go/wasi/io.go
package wasi

import "errors"

// BufferStream wraps a byte slice as an InputStream.  Used by tests
// and as a fallback when streaming data is already materialized.
type BufferStream struct {
	data []byte
	off  int
}

// NewBufferStream returns an InputStream backed by data.
func NewBufferStream(data []byte) *BufferStream {
	return &BufferStream{data: data}
}

// Read implements InputStream.
func (s *BufferStream) Read(buf []byte) (int, bool, error) {
	if s.off >= len(s.data) {
		return 0, true, nil
	}
	n := copy(buf, s.data[s.off:])
	s.off += n
	eof := s.off >= len(s.data)
	return n, eof, nil
}

// CaptureStream is an OutputStream that accumulates writes in memory.
// Used by tests to inspect produced bodies.
type CaptureStream struct {
	buf    []byte
	closed bool
}

// NewCaptureStream returns an empty CaptureStream.
func NewCaptureStream() *CaptureStream { return &CaptureStream{} }

// Bytes returns the accumulated body bytes (copy-free).
func (s *CaptureStream) Bytes() []byte { return s.buf }

// WriteAll implements OutputStream.
func (s *CaptureStream) WriteAll(chunk []byte) error {
	if s.closed {
		return errors.New("stream closed")
	}
	s.buf = append(s.buf, chunk...)
	return nil
}

// Close marks the stream closed.  Subsequent WriteAll calls fail.
func (s *CaptureStream) Close() error { s.closed = true; return nil }
